runq.c

```diff
@@ -38,23 +38,25 @@ typedef struct {
 
 typedef struct {
     // token embedding table
-    QuantizedTensor token_embedding_table; // (vocab_size, dim)
+    QuantizedTensor *q_tokens; // (vocab_size, dim)
+    float* token_embedding_table; // same, but dequantized
```

Comment:
is it better to just dequantize embeddings on demand as in my original pr? i don't super love that this way we're adding a lot of memory and latency

Reply:
(startup latency i mean, even though the per token latency would be improved by a tiny amount)

Reply:
Copying my comment from previous PR: 
> In this abstraction - QuantizedTensor is atomic, we do not index into or slice from it. We can matmul with it or dequantize all elements. This abstraction model breaks token embedding (sliced from it) & shared classifier weights (matmul-ed). So I split token embedding into separate rows and exported wcls as one quantizedtensor as well (as if it is not shared). 

So options I could think of are:
1) Partially dequantize QuantizedTensor and break current abstraction of it being opaque/atomic. 
2) Export each row/token embedding as separate QuantizedTensor - this turned out to be bit weird as we had to re-export wcls as 1 QuantizedTensor for final matmul, thus breaking weight sharing (This commit: https://github.com/karpathy/llama2.c/pull/383/commits/f850a97c6ad94b15b8d41fadfa9aa196d615306c)
3) Current approach in this PR - de quantize all at startup. I think this is the best of all options above.

Reply:
I'm not sure that I follow. Why can't we index into the QuantizedTensor into the correct row? This should work just fine as long as dim % GS == 0, which is the case via an assert in the python export code?

Reply:
We can index but goal was to avoid indexing to it from inside `forward()` function (if you want to add support for other quantization methods/techniques, i.e. int4). 
We can implement another `quantized_get_row()` or smth like that separately to decouple quantization from transformer/forward() code.

Reply:
Not sure if I appreciate the int4 issue but maybe I haven't stared at this enough and you're ahead of me. I think I'll merge this for now as it is quite nice.